File: Python/123.Best Time to Buy and Sell Stock III .py
#用一个数组表示股票每天的价格，数组的第i个数表示股票在第i天的价格。 如果最多允许两次交易，求最大的收益
#第一次的卖可以和第二次的买在同一时间，但第二次的买不能在第一次的卖左边
class Solution(object):
    def maxProfit(self, prices):
        l  = len(prices)
        if l<=1:
            return 0
        maxpro,curmin = 0,prices[0]
        premax , postmax = [0 for i in range(l)] , [0 for i in range(l)]
        for i in range(1,l):
            premax[i] = max(prices[i]-curmin,premax[i-1])
            curmin = min(curmin,prices[i])
        curmax = prices[l-1]
        i = l-2
        while i>0:
            postmax[i] = max(postmax[i+1],curmax-prices[i])
            curmax = max(curmax,prices[i])
            i -= 1
        for i in range(l):
            maxpro = max(maxpro,premax[i]+postmax[i])
        print(premax)
        print(postmax)
        return maxpro


s = Solution()
print(s.maxProfit([3,6,9,4,2,8,7,5]))

File: C++/123.Best Time to Buy and Sell Stock III.cpp
/*
最佳的购买和抛售股票的时机，最多买卖各两次，求最大收益。
而且同一天不能进行两次交易。
*/
/*
思路：动态规划。
 */

#include <iostream>
#include <vector>
#include <windows.h>
#include <windef.h>
#include <climits>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	int hold1 = INT_MIN, hold2 = INT_MIN;
        int release1 = 0, release2 = 0;
        for(int i=0; i < prices.size(); ++i)
        {   
        	// Assume we only have 0 money at first
            release2 = max(release2, hold2+prices[i]);     // The maximum if we've just sold 2nd stock so far.
            hold2    = max(hold2, release1-prices[i]);  // The maximum if we've just buy  2nd stock so far.
            release1 = max(release1, hold1+prices[i]);     // The maximum if we've just sold 1nd stock so far.
            hold1    = max(hold1, -prices[i]);          // The maximum if we've just buy  1st stock so far. 
        }
        return release2;
    }
};

int main()
{
	Solution s;
	vector<int> nums;
	nums.push_back(1);
	nums.push_back(2);
	nums.push_back(4);
	nums.push_back(2);
	nums.push_back(5);
	nums.push_back(7);
	nums.push_back(2);
	nums.push_back(4);
	nums.push_back(9);
	nums.push_back(0);
	cout<<s.maxProfit(nums)<<endl;
}